function-executor/src/wasmedge.rs

```diff
@@ -0,0 +1,855 @@
+//! Use WasmEdge as executor's runtime.
+
+#![warn(missing_docs)]
+
+use crate::{eyre_to_anyhow, kvs_get, kvs_put, EssaResult, FunctionExecutor};
+use anna::{
+    lattice::{LastWriterWinsLattice, Lattice},
+    nodes::ClientNode,
+    ClientKey,
+};
+use anyhow::Context;
+use essa_common::scheduler_function_call_topic;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use uuid::Uuid;
+use wasmedge_sdk::{
+    Executor, Func, ImportObject, ImportObjectBuilder, Instance, Memory, Module, Store,
+};
+use wasmedge_sys::types::WasmValue;
+use wasmedge_types::{ExternalInstanceType, ValType};
+use zenoh::{
+    prelude::{Receiver, Sample, SplitBuffer, ZFuture},
+    query::ReplyReceiver,
+    queryable::Query,
+};
+
+impl FunctionExecutor {
+    /// Runs the given WASM module.
+    pub fn run_module(mut self, wasm_bytes: Vec<u8>) -> anyhow::Result<()> {
+        log::info!("Start running WASM module");
+
+        // compile the WASM module
+        let module =
+            Arc::new(Module::from_bytes(None, &wasm_bytes).context("failed to load wasm module")?);
+
+        // store the compiled module in the key-value store under an
+        // unique key (to avoid recompiling the module on future function
+        // calls)
+        log::info!("Storing compiled wasm module in anna KVS");
+        let module_key: ClientKey = Uuid::new_v4().to_string().into();
+        let value = LastWriterWinsLattice::new_now(
+            // TODO: serialize the WasmEdge module into bytes, require WasmEdge core support.
+            wasm_bytes,
+        )
+        .into();
+        kvs_put(module_key.clone(), value, &mut self.anna)?;
+
+        let mut host_state = HostState {
+            module: module.clone(),
+            module_key,
+            function_result: None,
+            next_result_handle: 1,
+            results: HashMap::new(),
+            result_receivers: HashMap::new(),
+            zenoh: self.zenoh.clone(),
+            zenoh_prefix: self.zenoh_prefix.clone(),
+            anna: self.anna,
+        };
+
+        let mut instance_wrapper = InstanceWrapper::new()?;
+        instance_wrapper.register(&module, &mut host_state, vec![])?;
+
+        instance_wrapper.call_default()?;
+        Ok(())
+    }
+
+    /// Runs the given function of a already compiled WASM module.
+    pub fn handle_function_call(mut self, query: Query) -> anyhow::Result<()> {
```

Comment:
The first part of this method seems to be identical for `wasmedge` and `wasmtime`. Maybe we can deduplicate this somehow?